src/cascadia/TerminalControl/TermControl.cpp

```diff
@@ -743,6 +743,17 @@ namespace winrt::Microsoft::Terminal::Control::implementation
                     acrylic.BackgroundSource(Media::AcrylicBackgroundSource::HostBackdrop);
                 }

```

Comment:
I'd also deduplicate the function call like this:
```cpp
if (!acrylic)
{
    acrylic = Media::AcrylicBrush{};
}

const auto backdropStyle =
    _core.Settings().EnableUnfocusedAcrylic()
    ? Media::AcrylicBackgroundSource::Backdrop
    : Media::AcrylicBackgroundSource::HostBackdrop;
acrylic.BackgroundSource(backdropStyle);
```

It may also be worth it to first check `if (acrylic.BackgroundSource() != backdropStyle)`. At least to me it's unknown how well WinUI's `OnPropertyChanged` system works. Does it emit a change event if you call a setter with the property's current value? If it does, it may be worth the extra `if` condition, otherwise it doesn't matter.
(If unknown, I'd not use an `if` condition, because this feature is off by default.)

Reply:
> Could you just move this out of the `if (acrylic == nullptr)` instead of adding identical code in the `else` branch?

Thanks can't believe I missed that!

Reply:
@lhecker You replaced `acrylic == nulptr `with `!acrylic`. Is !acrylic better?  (New to c++)
Or are they like the same in c++?

Reply:
They're the same, and this codebase uses both. Personal style. :smile:

Reply:
I use `!` with objects because to me objects aren't pointers and so I don't compare them with `nullptr`. Basically, in my head, when I read `if (!object)`, I read "if not containing anything", whereas `if (object == nullptr)` to me reads more like "if my object equals the empty object", which is longer and doesn't quite match my thought process.

I do write `pointer == nullptr` quite often though, but in a different context. For instance when I write `return pointer == nullptr;` then that to me reads like "return if the pointer is empty" and that to me makes way more sense than the alternatives:
* `return pointer` ---> implicit conversion to `bool` = magic = opaque = bad
* `return !!pointer` ---> "return if not a not pointer" = tongue breaker = bad
* `return static_cast<bool>(pointer)` ---> my wrists hurt just from typing this out = bad

Reply:
@lhecker 
> If it was a regular C++ function call, the compiler could then better inline the call (since there's just 1 place it's called from). In case of a WinRT like this it's even more beneficial, because it cannot be inlined and cppwinrt generates imperfect assembly.
> 

I looked at it a bit and through this youtube video: https://www.youtube.com/watch?v=H6JvFtk_Fm0 and https://godbolt.org/ 

What I understand from what you are saying is that in cppwinrt the function can't be inlined so there would be 2 extra calls to assembly which would reduce the performance. Compared to 1 extra call with your suggested approach

![image](https://github.com/microsoft/terminal/assets/15957528/ac013f8e-38ea-4dc9-8794-bfb229d900ef)

Reply:
@lhecker 
> It may also be worth it to first check `if (acrylic.BackgroundSource() != backdropStyle)`. At least to me it's unknown how well WinUI's `OnPropertyChanged` system works. Does it emit a change event if you call a setter with the property's current value? If it does, it may be worth the extra `if` condition, otherwise it doesn't matter. (If it's unknown, I'd not use an extra `if` condition here.)

I'll look into it

Reply:
> What I understand from what you are saying is that in cppwinrt the function can't be inlined so there would be 2 extra calls to assembly which would reduce the performance. Compared to 1 extra call with your suggested approach right?
>
> Not sure what imperfect assembly means, what does that mean?

There's a slight difference between dynamic and static dispatch. Generally speaking, static dispatch is faster.

Static dispatch is when the compiler knows exactly what the definition of some function name is. That is, when you call `square` it knows that there's just one single `square` function and it'll call exactly that one. Since it knows which function it is, the compiler can guess much more accurately if it's worth it to inline the function call or to transform the function call in some way. For instance, if you have a function like this:
```cpp
void foo(bool toggle) {
    if (toggle) {
        // lots and lots of code here
    }
}
```
and you call it as `foo(false)`, then the compiler will realize that the entire function call is pointless since `toggle` is always `false` and the branch is never taken. It'll just completely erase it. (Side note: This only works for `static` functions in the same .c/.cpp file. For functions outside of your file, you need to use [LTCG](https://learn.microsoft.com/en-us/cpp/build/reference/ltcg-link-time-code-generation?view=msvc-170) which we do.)

Dynamic dispatch functions are called `virtual` in C++. Virtual functions are effectively function pointers. If you've never worked with `virtual` functions before, I think it's best if you look it up via Google. Other people can do a much better job of explaining this.
Anyways, the compiler can't generally know what function you call when you use dynamic dispatch, because there can be >1 implementation of any given interface. The exception is yet again when you use LTCG, _and_ if the compiler happens to be able to infer what concrete type your interface is. So if you have a `IControl` interface instance, but you created that one from a specific `Button` control instance then the compiler knows that your `Visible(true)` call can be statically dispatched to the `Button` class. Again, this only works if the compiler happens to see through your code. It'd be foolish to rely on that when performance is a concern.

WinRT uses something we could call dynamic^2 dispatch. WinRT fundamentally relies on interfaces for everything and has no static dispatch at all. It's based on COM and so it has this idea that if you have an instance of some object, you can ask that instance which other interfaces it fulfills (`QueryInterface`). This is effectively a form of dynamic dispatch as well (in pure C++ this is called `dynamic_cast`).
So when you construct a `Button` you actually construct an `IButton`. You can call all the `IButton` interface methods as if they were `virtual` functions in C++. They basically are.
But if you now want to call a method on the button runtimeclass that is not part of `IButton` but rather part of `IButton3` or whatever, then cppwinrt will first call `QueryInterface` to ask your `IButton` if it can please give a `IButton3` variant of itself, and only then you will call the actual function that you wanted to call (just an example).

A runtimeclass like `AcrylicBrush` for instance is actually 2 interfaces: `IAcrylicBrush` and `IAcrylicBrush2`. `IAcrylicBrush2` has just 2 methods: The getter and setter for `TintLuminosityOpacity`. So if you ever call `TintLuminosityOpacity` then the actual generated code looks like this:
```cpp
IAcrylicBrush* brush2;
brush2->QueryInterface(__uuidof(IAcrylicBrush2), &brush2); // dynamic dispatch 1
brush2->TintLuminosityOpacity(0.5);                        // dynamic dispatch 2
```

cppwinrt marks those `QueryInterface` calls as "pure" functions (= no side effects), which is a terrible lie, but it does get the compiler to make more aggressive optimizations. Code like that:
```cpp
for (...) {
    brush.TintLuminosityOpacity(0.5);
}
```
is transformed to:
```cpp
IAcrylicBrush* brush2;
brush2->QueryInterface(__uuidof(IAcrylicBrush2), &brush2);

for (...) {
    brush2->TintLuminosityOpacity(0.5);
}
```
...because the compiler believes that a side effect free function call can be "hoisted" out of loops. Again, it'd be foolish to rely on that when performance is a concern.

But there's one detail where WinRT differs from regular virtual function calls: Even if you use LTCG, you can be pretty certain that any WinRT function call is too difficult by any compiler to optimize. So if you write a WinRT class and it has a lot of small member functions that are called very frequently? That's 100% going to be a bottleneck.

---

In this particular code performance is absolutely no concern at all. WinUI as a UI framework is just not performant. Any such optimization we do will be absolutely dwarfed by whatever WinUI is doing internally, so we should write whatever is easiest.
I still think it's a good idea to do it "properly" if it's not too much of a burden, because certain performance problems are created by a million tiny cuts and this is one such tiny cut for sure. You can see the difference here when you scroll all the way down: https://godbolt.org/z/fc3h8xrf7

Since C++ templates can be very expressive, it can be very inobvious when you write really bad assembly. The `TintLuminosityOpacity` function mentioned above for instance, requires both a `QueryInterface` call and the argument isn't even a `double` - it's a `IReference<double>`! That is, calling that function actually requires the construction of an entire WinRT object instance. The difference between a branch and a ternary then becomes increasingly more obvious: https://godbolt.org/z/3GjnYs5Wc

Reply:
@lhecker 

Thanks for the thorough explanation. I get more lines of assembly is less performance. Some day I'll understand all of it. For now I think it suffices to know I can use this to compare performance between lines of code. 

Does this setting: `/std:c++20 /await /GR- /O2` simulate our environment with cpp winrt well?  Then I can use this in the future to test performance things.

Reply:
@lhecker 

About `if (acrylic.BackgroundSource() != backdropStyle)` I don't know.  Maybe I can look inside the WIndows::UI::Xaml and look for OnPropertyChanged and see what it does.

Reply:
> Does this setting: `/std:c++20 /await /GR- /O2` simulate our environment with cpp winrt well?  Then I can use this in the future to test performance things.

In short: Yes.
* `/std:c++20 /await` is required for cppwinrt to work
* `/GR-` disabled "RTTI" (= reflection). if you didn't do that the output of godbolt would be much larger and not be helpful for reading assembly
* `/O2` is basically Release mode optimizations

So if I don't use WinRT on godbolt I usually just use `/O2 /std:c++20 /GR-`.

> Then I can use this in the future to test performance things.

If you continue reading godbolt outputs, it'll definitely make sense over time, even if you never intended to learn assembly. Many instructions have an exhaustive description if you hover over them with your cursor. Strictly speaking, the amount of a small amount of instructions doesn't really correlate with performance, but that also will make sense over time. 😅

> Took a look at `if (acrylic.BackgroundSource() != backdropStyle)`.  Couldn't find any information about it in their documentation. Maybe I can look inside the WIndows::UI::Xaml and look for OnPropertyChanged and see what it does that might take me a while though probably better to leave it at this for now. 

Don't worry about it. I'm pretty sure it works alright even without it. 🙂

Reply:
> (In particular, there's not much reason to performance optimize UI code since almost none of it is "hot code". But knowing what's slow is still useful because it avoids making architectural designs that are inherently slow and can't be optimized in hindsight.)

What would you consider "hot code" in this project?